compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -1001,7 +1012,26 @@ class ExpressionCodegen(
 
     override fun consumeReifiedOperationMarker(typeParameter: TypeParameterMarker) {
         require(typeParameter is IrTypeParameterSymbol)
-        if (typeParameter.owner.parent != irFunction) {
+        // This is a hack to work around the problem in LocalDeclarationsLowering. Specifically, suppose an inline
+        // lambda uses a reified type parameter declared by a function:
+        //
+        //     object {
+        //         inline fun <reified T : Any> f() = run { T::class.java.getName() }
+        //     }
+        //
+        // LocalDeclarationsLowering would extract that lambda into a method of the enclosing type, but will not create
+        // a reified type parameter in it (in fact, the lambda method isn't even marked as inline):
+        //
+        //     object {
+        //         /* static */ private fun `f$lambda-0`() = T::class.java.getName()
+        //         inline fun <reified T : Any> f() = run(::`f$lambda-0`)
```

Comment:
@pyos Could you extract commit with this change ("JVM_IR: "fix" reified type parameter scope check" and linked one  "JVM_IR: generate SMAPs for classes local to inline lambdas...") to separate pr as I need more time for investigation.
Actually we have back logic in lowering made by @sfs  that merges f$lambda-0 fun into invoke method of lambda class/function reference, so maybe problem could be easier hacked here on class construction: https://github.com/JetBrains/kotlin/commit/c77fa1ecd6040be2dadc1d66e9013aa2aa49f594#diff-852213ef8c0292f5b7339ac0b276eab3R232

Reply:
Removed the commits from this branch. I'll create a separate PR later as there's a small conflict in ExpressionCodegen in a place that's also touchedby "JVM_IR: omit nullability assertions in anonymous object constructors".

Not sure what Steven's change has to do with this, as CallableReferenceLowering completely ignores inline lambdas. If you're planning on generating classes for them, the situation will become even worse as classes can't have reified type parameters.